data/summary/types/continuous.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/tidepool-org/platform/pointer"
+)
+
+type ContinuousBucketData struct {
+	TotalRecords int `json:"totalRecords" bson:"totalRecords"`
+
+	// RealtimeRecords is the total count of records which were both uploaded within 24h of the record creation
+	// and from a continuous dataset
+	RealtimeRecords int `json:"realtimeRecords" bson:"realtimeRecords"`
+
+	// DeferredRecords is the total count of records which are in continuous datasets, but not uploaded within 24h
+	DeferredRecords int `json:"deferredRecords" bson:"deferredRecords"`
+}
+
+type ContinuousPeriod struct {
+	TotalRecords *int `json:"totalRecords" bson:"totalRecords"`
+
+	AverageDailyRecords *float64 `json:"averageDailyRecords" bson:"averageDailyRecords"`
+
+	RealtimeRecords *int     `json:"realtimeRecords" bson:"realtimeRecords"`
+	RealtimePercent *float64 `json:"realtimeRecordsPercent" bson:"realtimeRecordsPercent"`
+	DeferredRecords *int     `json:"deferredRecords" bson:"deferredRecords"`
+	DeferredPercent *float64 `json:"deferredPercent" bson:"deferredPercent"`
+}
+
+type ContinuousPeriods map[string]*ContinuousPeriod
+
+type ContinuousStats struct {
+	Periods    ContinuousPeriods                                      `json:"periods" bson:"periods"`
+	Buckets    []*Bucket[*ContinuousBucketData, ContinuousBucketData] `json:"buckets" bson:"buckets"`
+	TotalHours int                                                    `json:"totalHours" bson:"totalHours"`
+}
+
+func (*ContinuousStats) GetType() string {
+	return SummaryTypeContinuous
+}
+
+func (*ContinuousStats) GetDeviceDataTypes() []string {
+	return DeviceDataTypes
+}
+
+func (s *ContinuousStats) Init() {
+	s.Buckets = make([]*Bucket[*ContinuousBucketData, ContinuousBucketData], 0)
+	s.Periods = make(map[string]*ContinuousPeriod)
+	s.TotalHours = 0
+}
+
+func (s *ContinuousStats) GetBucketsLen() int {
+	return len(s.Buckets)
+}
+
+func (s *ContinuousStats) GetBucketDate(i int) time.Time {
+	return s.Buckets[i].Date
+}
+
+func (s *ContinuousStats) ClearInvalidatedBuckets(earliestModified time.Time) (firstData time.Time) {
+	if len(s.Buckets) == 0 {
+		return
+	} else if earliestModified.After(s.Buckets[len(s.Buckets)-1].LastRecordTime) {
+		return s.Buckets[len(s.Buckets)-1].LastRecordTime
+	} else if earliestModified.Before(s.Buckets[0].Date) || earliestModified.Equal(s.Buckets[0].Date) {
+		// we are before all existing buckets, remake for GC
+		s.Buckets = make([]*Bucket[*ContinuousBucketData, ContinuousBucketData], 0)
+		return
+	}
+
+	offset := len(s.Buckets) - (int(s.Buckets[len(s.Buckets)-1].Date.Sub(earliestModified.UTC().Truncate(time.Hour)).Hours()) + 1)
+
+	for i := offset; i < len(s.Buckets); i++ {
+		s.Buckets[i] = nil
+	}
+	s.Buckets = s.Buckets[:offset]
+
+	if len(s.Buckets) > 0 {
+		return s.Buckets[len(s.Buckets)-1].LastRecordTime
+	}
+	return
+}
+
+func (s *ContinuousStats) Update(ctx context.Context, cursor DeviceDataCursor, dataRepo DeviceDataFetcher) error {
```

Comment:
The cursor here should only return data relevant for calculating the summary. You can further simplify this by creating a new struct which encapsulates the logic for filtering out data from non-continuous datasets.

```go
type DeviceDataCursor[T Datum] interface {
  HasNext(context.Context) bool
  GetNextData() []T, error
  RemainingBatchLength() int
}

type ContinuousDeviceDataCursor[T Datum] struct {
  cursor DeviceDataCursor
  dataRepo DataRepository
}

func (c *ContinuousDeviceDataCursor) GetNextData() ([]T, error) {
  // Check if each datum is continuous, filter out if not
}

func (c *ContinuousDeviceDataCursor) IsUploadContinuous(uploadId string) bool {
  // Cache it here
}
```